crates/nu-protocol/src/engine/pattern_match.rs

```diff
@@ -145,55 +145,13 @@ impl Matcher for Pattern {
-                                    Unit::Exabyte => {
-                                        *val == amount * 1000 * 1000 * 1000 * 1000 * 1000
-                                    }
-                                    Unit::Zettabyte => {
-                                        *val == amount * 1000 * 1000 * 1000 * 1000 * 1000 * 1000
-                                    }
-                                    Unit::Kibibyte => *val == amount * 1024,
-                                    Unit::Mebibyte => *val == amount * 1024 * 1024,
-                                    Unit::Gibibyte => *val == amount * 1024 * 1024 * 1024,
-                                    Unit::Pebibyte => *val == amount * 1024 * 1024 * 1024 * 1024,
-                                    Unit::Exbibyte => {
-                                        *val == amount * 1024 * 1024 * 1024 * 1024 * 1024
-                                    }
-                                    Unit::Zebibyte => {
-                                        *val == amount * 1024 * 1024 * 1024 * 1024 * 1024 * 1024
-                                    }
-                                    _ => false,
-                                },
-                                _ => false,
-                            }
-                        } else if let Value::Duration { val, .. } = &value {
-                            // FIXME: we probably want this math in one place that both the
-                            // pattern matcher and the eval engine can get to it
-                            match &amount.expr {
-                                Expr::Int(amount) => match &unit.item {
-                                    Unit::Nanosecond => val == amount,
-                                    Unit::Microsecond => *val == amount * 1000,
-                                    Unit::Millisecond => *val == amount * 1000 * 1000,
-                                    Unit::Second => *val == amount * 1000 * 1000 * 1000,
-                                    Unit::Minute => *val == amount * 1000 * 1000 * 1000 * 60,
-                                    Unit::Hour => *val == amount * 1000 * 1000 * 1000 * 60 * 60,
-                                    Unit::Day => *val == amount * 1000 * 1000 * 1000 * 60 * 60 * 24,
-                                    Unit::Week => {
-                                        *val == amount * 1000 * 1000 * 1000 * 60 * 60 * 24 * 7
-                                    }
-                                    _ => false,
-                                },
-                                _ => false,
-                            }
+                        let span = unit.span;
+
+                        if let Expr::Int(size) = amount.expr {
+                            &unit.item.to_value(size, span) == value
+                        } else if let Expr::Float(size) = amount.expr {
+                            let size = size.to_bits() as i64;
```

Comment:
that might be a very dumb question, but why do we need this here? :yum:

Reply:
wow, that is an amazingly detailed and clear answer :open_mouth: 

thank you very much for clarifying this @1Kinoti, really appreciate it :innocent:

Reply:
Reopening this because I am not sure if `.to_bits() as i64` is appropriate. This reinterpret-casts the float bit pattern but doesn't convert the Value in a lossy way to an integer with the same numeric value, right? 

Then the `Unit::to_value()` takes that integer and applies the transform for the unit based on integer arithmetic (as `i64` is the backing storage for both `Value::Filesize` and `Duration`

Reply:
i am not sure i understand your question here.

Reply:
@sholderbach, i went through [how values with unit are parsed](https://github.com/nushell/nushell/blob/9e9fe83bfd13433eddeb5fd92ecdefd100a9168f/crates/nu-parser/src/parser.rs#L2351) and i realized that the whole branch for floats is unnecessary because an `Expr::ValWithUnit` always ends up with an `i64` created [here](https://github.com/nushell/nushell/blob/9e9fe83bfd13433eddeb5fd92ecdefd100a9168f/crates/nu-parser/src/parser.rs#LL2392C1-L2401C11). i have removed it